rotary/rules/rotary.smk

```diff
@@ -312,16 +369,228 @@ rule qc_long_length_stats:
+    conda:
+        "../envs/qc.yaml"
+    log:
+        "{sample}/logs/qc/short/short_read_quality_trimming.log"
+    benchmark:
+        "{sample}/benchmarks/qc/short/short_read_quality_trimming.benchmark.txt"
+    params:
+        quality_trim_direction = config.get("quality_trim_direction"),
+        quality_trim_cutoff = config.get("quality_trim_score_cutoff"),
+        min_read_length = config.get("minimum_read_length_quality_trim"),
+        min_average_quality = config.get("minimum_average_quality_score_post_trim") if config.get("quality_trim_direction") != "f" else 0
+    threads:
+        config.get("threads", 1)
+    resources:
+        mem = config.get("memory")
+    shell:
+        """
+        bbduk.sh -Xmx{resources.mem}g threads={threads} in={input.short_r1} in2={input.short_r2} \
+          out={output.short_r1} out2={output.short_r2} qhist={output.quality_histogram} \
+          qtrim={params.quality_trim_direction} trimq={params.quality_trim_cutoff} minlength={params.min_read_length} \
+          minavgquality={params.min_average_quality} overwrite=t interleaved=f qin=33 \
+          2> {log}
+        """
+
+
+rule short_read_contamination_filter:
+    """
+    Filters short reads based on match to a reference. 
+    Note that contamination_references in the input is only there to trigger download of the genome files; the actual 
+    text to be input into the shell (with proper comma separation) is prepared in all_contaminant_references in params. 
+    The function get_contamination_reference_files called by this rule returns genome paths that can be understood by 
+    rule download_contamination_reference, which allows the download rule to be triggered as needed depending on which 
+    genomes the user wants to use as references. 
+    """
+    input:
+        short_r1 = "{sample}/qc/short/{sample}_quality_trim_R1.fastq.gz",
+        short_r2 = "{sample}/qc/short/{sample}_quality_trim_R2.fastq.gz",
+        contamination_references = get_contamination_reference_files(config.get("contamination_references"),
+            DB_DIR_PATH, PHIX_GENOME_ACCESSION, HUMAN_GENOME_ACCESSION)
+    output:
+        short_r1 = temp("{sample}/qc/short/{sample}_filter_R1.fastq.gz"),
+        short_r2 = temp("{sample}/qc/short/{sample}_filter_R2.fastq.gz"),
+        quality_histogram = "{sample}/qc/short/{sample}_filter_qhist.tsv",
+        filter_stats= "{sample}/stats/{sample}_short_read_contamination_filter.txt"
+    conda:
+        "../envs/qc.yaml"
+    log:
+        "{sample}/logs/qc/short/short_read_contamination_filter.log"
+    benchmark:
+        "{sample}/benchmarks/qc/short/short_read_contamination_filter.benchmark.txt"
+    params:
+        contamination_filter_kmer_length = config.get("contamination_filter_kmer_length"),
+        all_contaminant_references = ','.join(get_contamination_reference_files(config.get("contamination_references"),
+                DB_DIR_PATH, PHIX_GENOME_ACCESSION, HUMAN_GENOME_ACCESSION)) \
```

Comment:
@jmtsuji Move this to a function passed to params (I think that's possible?), or use this code to assign a global variable at the top of the file. This is too long of an inline if-else statement. Once the code is moved above, you can convert it to a traditional if-else statement using pycharm (right-click the if and select show context actions ). Also, you could assign the result of `get_contamination_reference_files(config.get("contamination_references"), DB_DIR_PATH, PHIX_GENOME_ACCESSION, HUMAN_GENOME_ACCESSION)` to the second global variable, which could be used in the prior mentioned if-else statement and could also be assigned to ```input.contamination_references```  for rule ```short_read_contamination_filter```

Reply:
@LeeBergstrand This variable was simplified so that an if-else statement is no longer needed. It's still a bit long, though (mostly because of long variable and function names) -- what do you think?

Reply:
Also, note that this will fail with a somewhat cryptic error message if the user provides bad input (e.g., a string, not a list) to `contamination_references_custom_filepaths` in the config. We could add a sanity check for this in the config check proposed in #79 . Otherwise, I could make a function to validate these input files.